Overhaul docs, harden tracelinks.c, expand tests, refresh toolchain (#8)
flox/tracelinks: README.md

# tracelinks

[![C/C++ CI](https://github.com/flox/tracelinks/actions/workflows/c-cpp.yml/badge.svg)](https://github.com/flox/tracelinks/actions/workflows/c-cpp.yml)

> Follow a path one component at a time and report every symbolic link along the way.

`tracelinks` is a small, dependency-free C utility that resolves a path the way
the kernel does — component by component — and prints each symbolic link it
crosses, the target the link points at, and the file type of the final
resolved object. Unlike `realpath`, which only prints the destination, or
`ls -l`, which only shows one link at a time, `tracelinks` shows you the
**whole chain** and flags the things that go wrong: dangling links, circular
links, and missing components.

It is developed and maintained by [Flox](https://flox.dev) and ships as part of

> the Flox CLI. Flox environments are assembled from symlinks into the Nix store

Author: ```suggestion
the Flox CLI installer. Flox environments are assembled from symlinks into the Nix store
```

Author: Done — "the Flox CLI installer" in a1948c8.

and synced through [FloxHub](https://hub.flox.dev), so understanding how a given
link resolves — and catching a broken or looping one — is a routine part of
working with and debugging them.

## Features

- **Traces the full link chain**, indenting each hop so nested links are easy to follow.
- **Detects dangling links** (a link whose target does not exist) and exits nonzero.
- **Detects circular links** and prints the cycle it found.
- **Reports the file type** of the final target (regular file, directory, device, FIFO, socket, …).
- **`--fstype`** annotates each reported path with the filesystem backing it.
- **`--absolute`** reports resolved paths as absolute paths.
- **`--keep-going`** continues to the next path after an error instead of stopping.
- **Cross-platform:** builds and runs on Linux and macOS with no third-party libraries.

## Installation

### With Flox

> `tracelinks` is bundled with the [Flox](https://flox.dev) CLI, so if you have

Author: ```suggestion
`tracelinks` is bundled with the [Flox](https://flox.dev) CLI installer, so if you have
```

Author: Done — applied in a1948c8.

> `tracelinks` is bundled with the [Flox](https://flox.dev) CLI, so if you have
> Flox installed it is already on your `PATH`.
> 
> ### From source

Author: Many people install Flox on a pre-existing Nix installation, and they can use Flox to install it to their environment of choice:

```suggestion

Alternatively, if you are using the Flox CLI installed on a pre-existing Nix installation, you can install it with `flox install tracelinks`.

### From source
```

Author: Added the `flox install tracelinks` note for pre-existing Nix installs in a1948c8.

### From source

The build needs only a C compiler and `make`; `help2man` is used to generate the
man page and `git` is used to fetch the software and determine its version string:

```console
$ git clone https://github.com/flox/tracelinks
$ cd tracelinks
$ make                    # builds ./tracelinks and ./tracelinks.1

> $ make test               # runs the test suite
> $ make install PREFIX=/usr/local

Author: Let's remove the mention of `make test` just in case the user is on an older MacOS with an incompatible `mktemp` version.

```suggestion
$ make                    # builds ./tracelinks and ./tracelinks.1
$ make install PREFIX=/usr/local
```

Author: Done in 8071e36 — dropped the `make test` line from the end-user "From source" instructions. I also added a dedicated `## Development` section covering the contributor flow: `flox activate`, `make` / `make test`, lint/format, and `flox build` with an example of running the resulting `./result-tracelinks/bin/tracelinks`. So the from-source path stays base-OS-friendly while testing lives in the Flox-based dev flow.

> 
> ```console
> $ git clone https://github.com/flox/tracelinks
> $ cd tracelinks
> $ flox activate           # provides gcc, gnumake, help2man, clang-tools, gdb
> $ make                    # builds ./tracelinks and ./tracelinks.1
> $ make test               # runs the test suite
> $ make install PREFIX=/usr/local
> ```
> 
> `make install` honors `PREFIX` (default `/usr/local`) and installs both the
> binary and its man page. Set `VERSION` to stamp the build, e.g.
> `make VERSION=1.0.1`.

Author: This does not work because the binary created with these instructions would have dynamic dependencies on libraries found in /nix/store, and these could be garbage collected over time and break the build in the process, e.g.
```
flox [tracelinks limeytexan/default (local)] brantley@ubuntu:~/src/tracelinks$ ldd tracelinks
        linux-vdso.so.1 (0x0000e6b7672a0000)
        libc.so.6 => /nix/store/1bdzriipn18184zpwh2lfdabk2gqcgxn-glibc-2.42-61/lib/libc.so.6 (0x0000e6b766e30000)
        /nix/store/1bdzriipn18184zpwh2lfdabk2gqcgxn-glibc-2.42-61/lib/ld-linux-aarch64.so.1 (0x0000e6b767250000)
flox [tracelinks limeytexan/default (local)] brantley@ubuntu:~/src/tracelinks$ objdump -p tracelinks | grep RUNPATH
  RUNPATH              /nix/store/1bdzriipn18184zpwh2lfdabk2gqcgxn-glibc-2.42-61/lib:/nix/store/7a3zsvzzm9ki2r3w2cl0mx12zjicinkj-gcc-15.2.0-lib/lib
flox [tracelinks limeytexan/default (local)] brantley@ubuntu:~/src/tracelinks$ 
```

It would be better to instead document the tools required for the build only, and advise people to install those tools using the tooling provided by their base O/S.

```suggestion
The build needs only a C compiler and `make`; `help2man` is used to generate the
man page and `git` is used to fetch the software and determine its version string:

```console
$ git clone https://github.com/flox/tracelinks
$ cd tracelinks
$ make                    # builds ./tracelinks and ./tracelinks.1
$ make test               # runs the test suite
$ make install PREFIX=/usr/local
```

`make install` honors `PREFIX` (default `/usr/local`) and installs both the
binary and its man page. Set `VERSION` to stamp the build, e.g.
`make VERSION=1.0.1`.
```

Author: Good catch on the /nix/store RUNPATH — applied your suggestion in a1948c8: the "From source" section now documents only the required tools (C compiler, `make`, `help2man`, `git`) and drops `flox activate`, so the built binary links against the base-OS libc rather than a GC-able store path.

To make the "`git` ... determine its version string" line accurate, I also wired the Makefile to derive `VERSION` from `git describe --tags --always --dirty`, falling back to `unknown` for archive/no-git builds (e.g. the pure Flox sandbox). `./tracelinks --version` now prints e.g. `v1.0.2-1-gf04d1de-dirty`.
